### PR TITLE
feat(gatsby): Memoize process.env access and repeated function calls

### PR DESCRIPTION
**Pull request by KyleAMathews**
As measured in the create-pages benchmark, this PR speeds up createPages by ~10%.

The `createPages` activity in the create-pages benchmark dropped from ~15s to 13.5s for 100k pages (i.e. when running `NUM_PAGES=100000 gatsby build`).

### CODE REVIEW

File: benchmarks/create-pages/gatsby-node.js [hunk 1]
```diff
@@ -1,10 +1,11 @@
 const NUM_PAGES = parseInt(process.env.NUM_PAGES || 5000, 10)
 
+const blankTemplate = require.resolve(`./src/templates/blank.js`)
```
Comment (KyleAMathews) [inline benchmarks/create-pages/gatsby-node.js:3]: This (not counted in 10% dropped) also took a second or two off the total time heh. Apparently there's a lot of things that start to take up human-noticable time if you call them 100k times.

File: packages/gatsby/src/redux/actions/public.js [hunk 2]
```diff
@@ -357,15 +359,15 @@ ${reservedFields.map(f => `  * "${f}"`).join(`\n`)}
...
-      pageComponentCache[originalPageComponent] = page.component
+      pageComponentCache.set(originalPageComponent, page.component)
     }
   }
 
   let internalComponentName
   if (page.path === `/`) {
     internalComponentName = `ComponentIndex`
   } else {
-    internalComponentName = `Component${pascalCase(page.path)}`
```
Comment (KyleAMathews) [inline packages/gatsby/src/redux/actions/public.js:368]: we were making this "pretty" for no good reason (it's not visible) and it just was burning CPU cycles.

File: packages/gatsby/src/redux/actions/public.js [hunk 3]
```diff
@@ -137,11 +140,10 @@ actions.deletePage = (page: PageInput) => {
   }
 }
 
-const pascalCase = _.flow(_.camelCase, _.upperFirst)
 const hasWarnedForPageComponentInvalidContext = new Set()
 const hasWarnedForPageComponentInvalidCasing = new Set()
 const hasErroredBecauseOfNodeValidation = new Set()
-const pageComponentCache = {}
```
Comment (KyleAMathews) [inline packages/gatsby/src/redux/actions/public.js:144]: this probably didn't matter... but I like sets more than objects 🤷

File: benchmarks/create-pages/gatsby-node.js (hunk 1 above)
Comment (LekoArts) [reply benchmarks/create-pages/gatsby-node.js:3]: Then we should double check that our docs use that pattern :)

File: packages/gatsby/src/redux/actions/public.js (hunk 2 above)
Comment (pieh) [reply packages/gatsby/src/redux/actions/public.js:368]: We actually never use it ourselves anymore. So I would suggest marking it as deprecated in https://github.com/gatsbyjs/gatsby/blob/172cf4d24ca6940839d0278db74e125cb7e8c682/packages/gatsby/src/schema/types/built-in-types.ts#L93
https://github.com/gatsbyjs/gatsby/blob/172cf4d24ca6940839d0278db74e125cb7e8c682/packages/gatsby/src/redux/types.ts#L28 add comment to remove this field completely in v4.

**Review by pieh — APPROVED** [review]

File: packages/gatsby/src/redux/actions/public.js (hunk 2 above)
Comment (KyleAMathews) [reply packages/gatsby/src/redux/actions/public.js:368]: Good idea, I'll do this in a follow-up PR.

File: benchmarks/create-pages/gatsby-node.js (hunk 1 above)
Comment (KyleAMathews) [reply benchmarks/create-pages/gatsby-node.js:3]: 👍

**Pull request merged**